avogadro/core/molecule.cpp

```diff
@@ -737,5 +847,203 @@ Residue& Molecule::residue(Index index)
+{
+  if (bondId < bondCount()) {
+    if (!m_graphDirty) {
+      m_graph.removeEdge(m_bondPairs[bondId].first, m_bondPairs[bondId].second);
+      m_graph.addEdge(pair.first, pair.second);
+    }
+    m_bondPairs[bondId] = pair;
+    return true;
+  }
+  return false;
+}
+
+unsigned char Molecule::bondOrder(Index bondId) const
+{
+  return bondId < m_bondOrders.size() ? m_bondOrders[bondId] : 0;
+}
+
+bool Molecule::setBondOrders(const Array<unsigned char>& orders)
+{
+  if (orders.size() == bondCount()) {
+    m_bondOrders = orders;
+    return true;
+  }
+  return false;
+}
+
+bool Molecule::setBondOrder(Index bondId, unsigned char order)
+{
+  if (bondId < bondCount()) {
+    m_bondOrders[bondId] = order;
+    return true;
+  }
+  return false;
+}
+
+Index Molecule::atomCount(unsigned char number) const
+{
+  Index count(0);
+  for (Array<unsigned char>::const_iterator it = m_atomicNumbers.begin();
+       it != m_atomicNumbers.end(); ++it) {
+    if (*it == number)
+      ++count;
+  }
+  return count;
+}
+
+bool Molecule::setAtomicNumbers(const Core::Array<unsigned char>& nums)
+{
+  if (nums.size() == atomCount()) {
+    m_atomicNumbers = nums;
+    return true;
+  }
+  return false;
+}
+
+bool Molecule::setAtomicNumber(Index atomId, unsigned char number)
+{
+  if (atomId < atomCount()) {
+    m_atomicNumbers[atomId] = number;
+    return true;
+  }
+  return false;
+}
+
+bool Molecule::hasCustomElements() const
+{
+  for (Array<unsigned char>::const_iterator it = m_atomicNumbers.begin(),
+                                            itEnd = m_atomicNumbers.end();
+       it != itEnd; ++it) {
+    if (Core::isCustomElement(*it))
+      return true;
+  }
+  return false;
+}
+
+std::map<unsigned char, size_t> Molecule::composition() const
+{
+  // A map of atomic symbols to their quantity.
+  std::map<unsigned char, size_t> composition;
+  for (Array<unsigned char>::const_iterator it = m_atomicNumbers.begin(),
+                                            itEnd = m_atomicNumbers.end();
+       it != itEnd; ++it) {
+    ++composition[*it];
+  }
+  return composition;
+}
+
+Index Molecule::findBond(Index atomId1, Index atomId2) const
+{
+  assert(atomId1 < atomCount());
+  assert(atomId2 < atomCount());
+
+  std::pair<Index, Index> pair = Molecule::makeBondPair(atomId1, atomId2);
+  Array<std::pair<Index, Index>>::const_iterator iter =
+    std::find(m_bondPairs.begin(), m_bondPairs.end(), pair);
+
+  Index index = static_cast<Index>(std::distance(m_bondPairs.begin(), iter));
+  return index;
+}
+
+bool Molecule::removeBonds(Index atom)
+{
+  if (atom >= bondCount())
+    return false;
+  Index i = 0;
+  while (i < m_bondPairs.size()) {
+    auto& bond = m_bondPairs[i];
+    if (bond.first == atom || bond.second == atom) {
+      removeBond(i);
+    } else {
+      ++i;
+    }
+  }
+  return true;
+}
+
+Array<std::pair<Index, Index>> Molecule::getBonds(Index index) const
```

Comment:
This could probably use Molecule::getAtomBonds(index) to clarify it's  the  index to the atom, not a bond index.